api/handlers_api.go

```diff
@@ -110,42 +89,76 @@ func (r *Request) HandleJoinPlayer() (*mc.Message[mc.RespJoinGame], *mb.Game) {
 	r.Session.Player = joinPlayer
 
 	resp.AddPayload(mc.RespJoinGame{GameUuid: game.Uuid, PlayerUuid: joinPlayer.Uuid})
-	return &resp, game
+	return resp, game
+}
+
+// User will choose the configurations of ships on defence grid.
+// Then the grid is sent to backend and adjustment happens accordingly.
+func (r *Request) HandleReadyPlayer() (mc.Message[mc.NoPayload], *mb.Game) {
+	var readyPlayerReq mc.Message[mc.ReqReadyPlayer]
+	resp := mc.NewMessage[mc.NoPayload](mc.CodeReady)
+
+	if err := json.Unmarshal(r.Payload, &readyPlayerReq); err != nil {
+		resp.AddError(err.Error(), cerr.ConstErrInvalidPayload)
+		return resp, nil
+	}
+
+	game, player, err := r.Session.GameManager.FindGameAndPlayer(readyPlayerReq.Payload.GameUuid, readyPlayerReq.Payload.PlayerUuid)
+	if err != nil {
+		resp.AddError(err.Error(), cerr.ConstErrReady)
+		return resp, nil
+	}
+
+	// Check to see if rows and cols are equal to game's grid size
+	rows := len(readyPlayerReq.Payload.DefenceGrid)
+	if rows != game.GridSize {
+		resp.AddError(cerr.ErrDefenceGridRowsOutOfBounds(rows, game.GridSize).Error(), cerr.ConstErrReady)
+		return resp, nil
+	}
+	cols := len(readyPlayerReq.Payload.DefenceGrid[0])
+	if cols != game.GridSize {
+		resp.AddError(cerr.ErrDefenceGridColsOutOfBounds(cols, game.GridSize).Error(), cerr.ConstErrReady)
+		return resp, nil
+	}
+
+	player.SetDefenceGrid(readyPlayerReq.Payload.DefenceGrid)
+	player.IsReady = true
+	return resp, game
 }
 
 // Handle the attack logic for the incoming request
-func (r *Request) HandleAttack() (*mc.Message[mc.RespAttack], *mb.Player) {
+func (r *Request) HandleAttack() (mc.Message[mc.RespAttack], *mb.Player) {
 	var reqAttack mc.Message[mc.ReqAttack]
 	resp := mc.NewMessage[mc.RespAttack](mc.CodeAttack)
 
 	if err := json.Unmarshal(r.Payload, &reqAttack); err != nil {
 		resp.AddError(err.Error(), cerr.ConstErrInvalidPayload)
-		return &resp, nil
-	}
-
-	x := reqAttack.Payload.X
-	y := reqAttack.Payload.Y
-	if x > mb.GameValidUpperBound || y > mb.GameValidUpperBound || x < mb.GameValidLowerBound || y < mb.GameValidLowerBound {
-		resp.AddError(cerr.ErrXorYOutOfGridBound(x, y).Error(), cerr.ConstErrAttack)
-		return &resp, nil
+		return resp, nil
 	}
 
 	game, attacker, err := r.Session.GameManager.FindGameAndPlayer(reqAttack.Payload.GameUuid, reqAttack.Payload.PlayerUuid)
 	if err != nil {
 		resp.AddError(err.Error(), cerr.ConstErrAttack)
-		return &resp, nil
+		return resp, nil
+	}
+
+	x := reqAttack.Payload.X
+	y := reqAttack.Payload.Y
+	if x > game.ValidUpperBound || y > game.ValidUpperBound || x < game.ValidLowerBound || y < game.ValidLowerBound {
```

Comment:
Since we have a `Coordinate` struct now, you can make `coordinate` here and have a function in game to validate that coordinate, something like:
```go
coordinate = NewCoordinate(x: reqAttack.Payload.x, y: reqAttack.Payload.y)
if game.isCoordinateValid(coordinate) {  }
```

Reply:
I really like this yes